Updated Item property override in 1.16

PellucidDice opened the pull request.

PellucidDice committed:
Update Item Properties for 1.16.x

PellucidDice committed:
Update overrides.md

TeamSpen210 commented on docs/models/overrides.md:
```diff
@@ -37,15 +39,18 @@ The format of an override can be seen on the [wiki][format], and a good example
-  public float apply(ItemStack stack, @Nullable World world, @Nullable EntityLivingBase entity) {
-    return (float)getPowerLevel(stack) / (float)getMaxPower(stack); // Some external methods
-  }
+private void setup(final FMLCommonSetupEvent event)
```
Shouldn't this be `FMLClientSetupEvent`?
```suggestion
private void setup(final FMLClientSetupEvent event)
```

PellucidDice committed:
Apply suggestions from code review

Changed FMLCommonSetupEvent to FMLClientSetupEvent

Co-authored-by: Spencer Brown <TeamSpen210@users.noreply.github.com>

PellucidDice committed:
Update method names

func_239420_a_ -> registerGlobalProperty
func_239418_a_ -> registerProperty

desht commented on docs/models/overrides.md:
```diff
@@ -1,14 +1,16 @@
-`Item::addPropertyOverride` is used to add a property to an item. The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a function that takes the `ItemStack`, the `World` it's in, and the `EntityLivingBase` that holds it, returning the `float` value for the property. Some examples are the `"pulling"` and "`pull`" pro ...
+`ItemModelsProperties::registerGlobalProperty` is used to add a property to all items, it does not take `Item` as it's parameter. `ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation ...
+!!! important
+    Use DeferredWorkQueue to proceed the task, since the data structures in ItemModelsProperties are not threadsafe.
```
`DeferredWorkQueue` has been replaced by `event.enqueueWork()`

desht commented on docs/models/overrides.md:
```diff
@@ -37,15 +39,18 @@ The format of an override can be seen on the [wiki][format], and a good example
-  }
+private void setup(final FMLClientSetupEvent event)
+{
+  DeferredWorkQueue.runLater(() ->
```
As above; use `event.enqueueWork()` here

desht commented on docs/models/overrides.md:
```diff
@@ -37,15 +39,18 @@ The format of an override can be seen on the [wiki][format], and a good example
+{
+  DeferredWorkQueue.runLater(() ->
+  {
+    ItemModelsProperties.func_239418_a_(ExampleItems.APPLE, 
```
this is `registerProperty()` in recent mappings

PellucidDice committed:
Update to fit 1.16.4

- Changed DeferredWorkQueue to ParallelDispatchEvent#enqueueWork
- Use the mapped name in the example from unmapped

PellucidDice commented:
Thank you for pointing these out.

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -1,14 +1,16 @@
+Item properties are a way for the "properties" of items to be exposed to the model system. An example is the bow, where the most important property is how far the bow has been pulled. This information is then used to choose a model for the bow, creating an animation for pulling it.
 
-An item property assigns a certain `float` value to every `ItemStack` it is registered for, and vanilla item model definitions can use these values to define "overrides", where an item defaults to a certain model, but if an override matches, it overrides the model and uses another. The format of item models, including overrides, can be found on the [wiki][]. They are useful mainly because of the  ...
+An item property assigns a certain `float` value to every `ItemStack` it is registered for, and vanilla item model definitions can use these values to define "overrides", where an item defaults to a certain model, but if an override matches, it overrides the model and uses another. They are useful mainly because they are continuous. For example, bows use item properties to define their pull anima ...
```
Could you add some clarification/explicit info. that the `float` value is not limited to `0.0`-`1.0`, and can be any value in the range of a `float`?
> Since the value of the property is a `float`, it increases continuously from 0 to 1.

This implies that the value is restricted to `0.0`-`1.0`, which can be confusing when encountering this for the first time.

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -1,14 +1,16 @@
 --------------------------
 
-`Item::addPropertyOverride` is used to add a property to an item. The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a function that takes the `ItemStack`, the `World` it's in, and the `EntityLivingBase` that holds it, returning the `float` value for the property. Some examples are the `"pulling"` and "`pull`" pro ...
+`ItemModelsProperties::registerGlobalProperty` is used to add a property to all items, it does not take `Item` as it's parameter. `ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation ...
```
> `ItemModelsProperties::registerGlobalProperty` is used to add a property to all items, it does not take `Item` as it's parameter. 

The method is by default `private static`. Should either be removed, or moved to the bottom and clarified that it is `private`.

> The `IItemPropertyGetter` is a function that **...**
It's more of a functional interface. Might want to mention that too.

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -37,15 +39,18 @@ The format of an override can be seen on the [wiki][format], and a good example
+  event.enqueueWork(() ->
+  {
+    ItemModelsProperties.registerProperty(ExampleItems.APPLE, 
+      new ResourceLocation(ExampleMod.MODID, pulling"), (stack, world, living) -> {
```
> ```java
>     new ResourceLocation(ExampleMod.MODID, pulling"), (stack, world, living) -> {
> ```
Missing starting `"` for the string .

sciwhiz12 reviewed (changes requested):
A few changes are in order, then this'll be ready.

If you have the time too, could you also add to the `Using Overrides` section that if multiple overrides match, the one nearest to the end of the list is used.

PellucidDice committed:
Update overrides.md

PellucidDice committed:
Update overrides.md

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,22 @@
 
-`Item::addPropertyOverride` is used to add a property to an item. The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a function that takes the `ItemStack`, the `World` it's in, and the `EntityLivingBase` that holds it, returning the `float` value for the property. Some examples are the `"pulling"` and "`pull`" pro ...
+`ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a functional interface that takes the `ItemStack`, the `ClientWorld` it's in (may be null ...
+There's also another private method `ItemModelsProperties::registerGlobalProperty` that is used to add properties to all items, and it does not take `Item` as it's parameter since all items will apply this property.
```
Minor nitpick: 
```suggestion
There's also another private method `ItemModelsProperties::registerGlobalProperty` that is used to add properties to all items, and it does not take `Item` as its parameter since all items will apply this property.
```

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,22 @@
 --------------------------
 
-`Item::addPropertyOverride` is used to add a property to an item. The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a function that takes the `ItemStack`, the `World` it's in, and the `EntityLivingBase` that holds it, returning the `float` value for the property. Some examples are the `"pulling"` and "`pull`" pro ...
+`ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a functional interface that takes the `ItemStack`, the `ClientWorld` it's in (may be null ...
```
Minor nitpick:
```suggestion
`ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a functional interface that takes the `ItemStack`, the `ClientWorld` it's in (may be null), and the `LivingEntity` that holds it (may be null), returning the `float` value for the property. For modded item properties, it is recommended that the modid of the mod is used as the namespace (e.g. `examplemod:property` and not just `property`, as that really means `minecraft:property`). These should be done in FMLClientSetupEvent.
```

sciwhiz12 reviewed (changes requested):
Just two minor nitpicky changes, then it's okay to me.

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

PellucidDice commented:
Thank you for reviewing.

sciwhiz12 reviewed (dismissed):
I will note that #325 does touch the same file, so there's a possibility that we'll need to have this rebased if that PR gets merged in first, if there's a merge conflict.

tterrag1098 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,22 @@
 ---------------
 
-The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model.
+The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model, but if there are multiple matches the predicate that is nearest to the end of the list will be used.
```
This addition is a bit wordy.
```suggestion
The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model, but if there are multiple matches last match in the list will be selected.
```

tterrag1098 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,22 @@
+Item properties are a way for the "properties" of items to be exposed to the model system. An example is the bow, where the most important property is how far the bow has been pulled. This information is then used to choose a model for the bow, creating an animation for pulling it.
 
-An item property assigns a certain `float` value to every `ItemStack` it is registered for, and vanilla item model definitions can use these values to define "overrides", where an item defaults to a certain model, but if an override matches, it overrides the model and uses another. The format of item models, including overrides, can be found on the [wiki][]. They are useful mainly because of the  ...
+An item property assigns a certain `float` value to every `ItemStack` it is registered for, and vanilla item model definitions can use these values to define "overrides", where an item defaults to a certain model, but if an override matches, it overrides the model and uses another. They are useful mainly because they are continuous. For example, bows use item properties to define their pull anima ...
```
In *this* case? Which case? This should probably just say "is not limited but generally between..."
```suggestion
An item property assigns a certain `float` value to every `ItemStack` it is registered for, and vanilla item model definitions can use these values to define "overrides", where an item defaults to a certain model, but if an override matches, it overrides the model and uses another. They are useful mainly because they are continuous. For example, bows use item properties to define their pull animation. The item models are decided by the 'float' number predicates, it is not limited but generally between `0.0F` and `1.0F`. This allows resource packs to add as many models as they want for the bow pulling animation along that spectrum, instead of being stuck with four "slots" for their models in the animation. The same is true of the compass and clock.
```

PellucidDice replied:
It's using vanilla bow as the example.

tterrag1098 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,22 @@
-`Item::addPropertyOverride` is used to add a property to an item. The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a function that takes the `ItemStack`, the `World` it's in, and the `EntityLivingBase` that holds it, returning the `float` value for the property. Some examples are the `"pulling"` and "`pull`" pro ...
+`ItemModelsProperties::registerProperty` is used to add a property to a certain item. The `Item` parameter is the item the property is attaching to (e.g. ExampleItems.APPLE). The `ResourceLocation` parameter is the name given to the property (e.g. `new ResourceLocation("pull")`). The `IItemPropertyGetter` is a functional interface that takes the `ItemStack`, the `ClientWorld` it's in (may be null ...
+There's also another private method `ItemModelsProperties::registerGlobalProperty` that is used to add properties to all items, and it does not take `Item` as its parameter since all items will apply this property.
+!!! important
```
Blank line before an admonition

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: tterrag <tterrag1098@gmail.com>

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: tterrag <tterrag1098@gmail.com>

PellucidDice committed:
Update overrides.md

sciwhiz12 commented on docs/models/overrides.md:
```diff
@@ -1,19 +1,23 @@
 ---------------
 
-The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model.
+The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model, but if there are multiple matches last match in the list will be selected.
```
Minor adjustment:
```suggestion
The format of an override can be seen on the [wiki][format], and a good example can be found in `model/item/bow.json`. For reference, here is a hypothetical example of an item with an `examplemod:power` property. If the values have no match, the default is the current model, but if there are multiple matches, the last match in the list will be selected.
```

sciwhiz12 commented on docs/models/overrides.md:
````diff
@@ -37,15 +41,18 @@ The format of an override can be seen on the [wiki][format], and a good example
 ```
 
-And here's a hypothetical snippet from the supporting code. (This does not have to be client-only; it will work on a server too. In vanilla, properties are registered in the item's constructor.)
+And here's a hypothetical snippet from the supporting code. Unlike the older versions (lower than 1.16.x), this needs to be done on client side only because ItemModelsProperties does not exist on server.
````
Another minor formatting adjustment:
```suggestion
And here's a hypothetical snippet from the supporting code. Unlike the older versions (lower than 1.16.x), this needs to be done on client side only because `ItemModelsProperties` does not exist on server.
```

sciwhiz12 reviewed (changes requested):
Just two more minor formatting changes, then it should be fine now.

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

PellucidDice committed:
Update docs/models/overrides.md

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

PellucidDice commented:
How would I go about with the conflicts?

ChampionAsh5357 commented:
You can rebase your branch on top of the 1.16.x upstream or merge the 1.16.x upstream with your current branch. Essentially, the file you are using is outdated as a current update to the branch modified the file in some way. If you happen to modify a portion of the file that was changed, you get a conflict. This will explain how to do this within the [command line](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-using-the-command-line). If you're using Github Desktop, it should pop up with a notification when trying to rebase/merge that will allow you to resolve the files in your selected text editor.

PellucidDice commented:
Sorry for my unclear question, what I meant to ask was which content you think is better/easier for the users? I'm not quite sure should I override the conflict section with his/her content, or override with mine.

ChampionAsh5357 commented:
#324 was written back for base 1.16 with relatively few changes to the source and rather updating method calls that were used within the new system. As such, it did not aim to expand or clarify some details, just to make the information being presented accurate. Since your content expands upon the information already available and clarifies some resources in the example source, I would say overriding the current changes is fine.

PellucidDice committed:
Merge branch '1.16.x' into item-property-override
